src/rcc.rs

```diff
@@ -75,20 +109,77 @@ impl CFGR {
         self
     }
 
+    pub fn enable_hsi(mut self, is_enabled: bool) -> Self {
+        self.enable_hsi = Some(is_enabled);
+        self
+    }
+
+    pub fn enable_hsi14(mut self, is_enabled: bool) -> Self {
+        self.enable_hsi14 = Some(is_enabled);
+        self
+    }
+
+    #[cfg(any(feature = "stm32f042", feature = "stm32f072"))]
+    pub fn enable_hsi48(mut self, is_enabled: bool) -> Self {
+        self.enable_hsi48 = Some(is_enabled);
+        self
+    }
+
+    pub fn enable_lsi(mut self, is_enabled: bool) -> Self {
+        self.enable_lsi = Some(is_enabled);
+        self
+    }
+
+    pub fn enable_pll(mut self, is_enabled: bool) -> Self {
+        self.enable_pll = Some(is_enabled);
+        self
+    }
+
     pub fn freeze(self) -> Clocks {
-        let pllmul = (4 * self.sysclk.unwrap_or(HSI) + HSI) / HSI / 2;
-        let pllmul = core::cmp::min(core::cmp::max(pllmul, 2), 16);
-        let sysclk = pllmul * HSI / 2;
+        // Default to lowest frequency clock on all systems.
+        let sysclk = self.sysclk.unwrap_or(HSI);
 
-        let pllmul_bits = if pllmul == 2 {
-            None
+        let r_sysclk; // The "real" sysclock value, calculated below
+        let src_clk_freq; // Frequency of source clock for PLL and etc, HSI, or HSI48 on supported systems.
+        let pllmul_bits;
+
+        // Select clock source based on user input and capability
+        // Highest selected frequency source available takes precident.
+        // For F04x, F07x, F09x parts, use HSI48 if requested.
+        if self.enable_hsi48.unwrap_or(false) { 
+            // This does not work... #[cfg(any(feature = "stm32f042", feature = "stm32f072"))]
+            src_clk_freq = HSI48; // Use HSI48 if requested and available.
+        } else if self.enable_hsi.unwrap_or(true) {
+            src_clk_freq = HSI; // HSI if requested, or by default.
+        } else {
+            src_clk_freq = HSI; // If no clock source is selclected use HSI.
+        }
+       
+        // Pll check
+        let enable_pll;
+        if sysclk == src_clk_freq {
+            // bypass pll if src clk and requested sysclk are the same, to save power.
+            // The only reason to override this behaviour is if the sysclk source were HSI, and you
+            // were running the USB off the PLL...
+            enable_pll = false;
+            pllmul_bits = None;
+            r_sysclk = src_clk_freq;
         } else {
-            Some(pllmul as u8 - 2)
-        };
+            let pllmul = (4 * self.sysclk.unwrap_or(src_clk_freq) + src_clk_freq) / src_clk_freq / 2;
+            let pllmul = cmp::min(cmp::max(pllmul, 2), 16);
+            r_sysclk = pllmul * src_clk_freq / 2;
+
+            pllmul_bits = if pllmul == 2 {
+                None
+            } else {
+                Some(pllmul as u8 - 2)
+            };
+            enable_pll = true; // Force PLL on.
```

Comment:
I think I am going to remove this as well.  enable_pll isn't needed because pllmul_bits acts as the gate for enabling the PLL...

Suggestions?

Reply:
Yeah, I'd rename pllmul_bits to pllmul and as that's already Option it's alright